File: chinese-family-tree/frontend/lib/models/person.dart
class Person {
  final int? id;
  final String uuid;
  final String familyName;
  final String givenName;
  final String? generationName;
  final String? courtesyName;
  final String? artName;
  final String? englishName;
  final String gender;
  final String? birthDate;
  final String? birthDateLunar;
  final String? birthPlace;
  final String? deathDate;
  final String? deathPlace;
  final String? burialPlace;
  final bool isDeceased;
  final bool isAdopted;
  final String? biography;
  final String? achievements;
  final String? occupation;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Person({
    this.id,
    required this.uuid,
    required this.familyName,
    required this.givenName,
    this.generationName,
    this.courtesyName,
    this.artName,
    this.englishName,
    required this.gender,
    this.birthDate,
    this.birthDateLunar,
    this.birthPlace,
    this.deathDate,
    this.deathPlace,
    this.burialPlace,
    this.isDeceased = false,
    this.isAdopted = false,
    this.biography,
    this.achievements,
    this.occupation,
    this.createdAt,
    this.updatedAt,
  });

  /// Full Chinese name (姓 + 名)
  String get fullName => '$familyName$givenName';

  /// Display name with generation name if available
  String get displayName {
    if (generationName != null && generationName!.isNotEmpty) {
      return '$familyName$generationName$givenName';
    }
    return fullName;
  }

  /// Formatted gender in Chinese
  String get genderChinese => gender == 'M' ? '男' : gender == 'F' ? '女' : '未知';

  /// Age calculation (approximate)
  int? get age {
    if (birthDate == null) return null;
    try {
      final birth = DateTime.parse(birthDate!);
      final now = DateTime.now();
      int age = now.year - birth.year;
      if (now.month < birth.month ||
          (now.month == birth.month && now.day < birth.day)) {
        age--;
      }
      return age;
    } catch (e) {
      return null;
    }
  }

  /// Whether the person is alive
  bool get isAlive => !isDeceased;

  /// Whether the person is the heir (eldest son)
  bool get isHeir => false; // Default to false, can be extended later

  factory Person.fromJson(Map<String, dynamic> json) {
    return Person(
      id: json['id'] as int?,
      uuid: json['uuid'] as String,
      familyName: json['family_name'] as String,
      givenName: json['given_name'] as String,
      generationName: json['generation_name'] as String?,
      courtesyName: json['courtesy_name'] as String?,
      artName: json['art_name'] as String?,
      englishName: json['english_name'] as String?,
      gender: json['gender'] as String,
      birthDate: json['birth_date'] as String?,
      birthDateLunar: json['birth_date_lunar'] as String?,
      birthPlace: json['birth_place'] as String?,
      deathDate: json['death_date'] as String?,
      deathPlace: json['death_place'] as String?,
      burialPlace: json['burial_place'] as String?,
      isDeceased: json['is_deceased'] as bool? ?? false,
      isAdopted: json['is_adopted'] as bool? ?? false,
      biography: json['biography'] as String?,
      achievements: json['achievements'] as String?,
      occupation: json['occupation'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'uuid': uuid,
      'family_name': familyName,
      'given_name': givenName,
      if (generationName != null) 'generation_name': generationName,
      if (courtesyName != null) 'courtesy_name': courtesyName,
      if (artName != null) 'art_name': artName,
      if (englishName != null) 'english_name': englishName,
      'gender': gender,
      if (birthDate != null) 'birth_date': birthDate,
      if (birthDateLunar != null) 'birth_date_lunar': birthDateLunar,
      if (birthPlace != null) 'birth_place': birthPlace,
      if (deathDate != null) 'death_date': deathDate,
      if (deathPlace != null) 'death_place': deathPlace,
      if (burialPlace != null) 'burial_place': burialPlace,
      'is_deceased': isDeceased,
      'is_adopted': isAdopted,
      if (biography != null) 'biography': biography,
      if (achievements != null) 'achievements': achievements,
      if (occupation != null) 'occupation': occupation,
    };
  }

  Person copyWith({
    int? id,
    String? uuid,
    String? familyName,
    String? givenName,
    String? generationName,
    String? courtesyName,
    String? artName,
    String? englishName,
    String? gender,
    String? birthDate,
    String? birthDateLunar,
    String? birthPlace,
    String? deathDate,
    String? deathPlace,
    String? burialPlace,
    bool? isDeceased,
    bool? isAdopted,
    String? biography,
    String? achievements,
    String? occupation,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Person(
      id: id ?? this.id,
      uuid: uuid ?? this.uuid,
      familyName: familyName ?? this.familyName,
      givenName: givenName ?? this.givenName,
      generationName: generationName ?? this.generationName,
      courtesyName: courtesyName ?? this.courtesyName,
      artName: artName ?? this.artName,
      englishName: englishName ?? this.englishName,
      gender: gender ?? this.gender,
      birthDate: birthDate ?? this.birthDate,
      birthDateLunar: birthDateLunar ?? this.birthDateLunar,
      birthPlace: birthPlace ?? this.birthPlace,
      deathDate: deathDate ?? this.deathDate,
      deathPlace: deathPlace ?? this.deathPlace,
      burialPlace: burialPlace ?? this.burialPlace,
      isDeceased: isDeceased ?? this.isDeceased,
      isAdopted: isAdopted ?? this.isAdopted,
      biography: biography ?? this.biography,
      achievements: achievements ?? this.achievements,
      occupation: occupation ?? this.occupation,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: chinese-family-tree/backend/internal/models/person.go
package models

import "time"

// Person represents an individual in the family tree
type Person struct {
	ID              int64      `db:"id" json:"id"`
	UUID            string     `db:"uuid" json:"uuid"`
	FamilyName      string     `db:"family_name" json:"family_name"`      // 姓
	GivenName       string     `db:"given_name" json:"given_name"`        // 名
	FullName        string     `db:"full_name" json:"full_name"`          // 姓名 (computed)
	GenerationName  *string    `db:"generation_name" json:"generation_name,omitempty"` // 字辈
	CourtesyName    *string    `db:"courtesy_name" json:"courtesy_name,omitempty"`     // 字
	ArtName         *string    `db:"art_name" json:"art_name,omitempty"`               // 号
	EnglishName     *string    `db:"english_name" json:"english_name,omitempty"`
	Gender          string     `db:"gender" json:"gender"`                // M, F, U
	BirthDate       *string    `db:"birth_date" json:"birth_date,omitempty"`
	BirthDateLunar  *string    `db:"birth_date_lunar" json:"birth_date_lunar,omitempty"` // 农历
	BirthPlace      *string    `db:"birth_place" json:"birth_place,omitempty"`
	DeathDate       *string    `db:"death_date" json:"death_date,omitempty"`
	DeathPlace      *string    `db:"death_place" json:"death_place,omitempty"`
	BurialPlace     *string    `db:"burial_place" json:"burial_place,omitempty"` // 墓地
	IsDeceased      bool       `db:"is_deceased" json:"is_deceased"`
	IsAdopted       bool       `db:"is_adopted" json:"is_adopted"`
	Biography       *string    `db:"biography" json:"biography,omitempty"`
	Achievements    *string    `db:"achievements" json:"achievements,omitempty"`
	Occupation      *string    `db:"occupation" json:"occupation,omitempty"`
	CreatedAt       time.Time  `db:"created_at" json:"created_at"`
	UpdatedAt       time.Time  `db:"updated_at" json:"updated_at"`
}

// CreatePersonRequest represents the request to create a person
type CreatePersonRequest struct {
	FamilyName      string  `json:"family_name" binding:"required"`
	GivenName       string  `json:"given_name" binding:"required"`
	GenerationName  *string `json:"generation_name,omitempty"`
	CourtesyName    *string `json:"courtesy_name,omitempty"`
	ArtName         *string `json:"art_name,omitempty"`
	EnglishName     *string `json:"english_name,omitempty"`
	Gender          string  `json:"gender" binding:"required,oneof=M F U"`
	BirthDate       *string `json:"birth_date,omitempty"`
	BirthDateLunar  *string `json:"birth_date_lunar,omitempty"`
	BirthPlace      *string `json:"birth_place,omitempty"`
	DeathDate       *string `json:"death_date,omitempty"`
	DeathPlace      *string `json:"death_place,omitempty"`
	BurialPlace     *string `json:"burial_place,omitempty"`
	IsDeceased      bool    `json:"is_deceased"`
	IsAdopted       bool    `json:"is_adopted"`
	Biography       *string `json:"biography,omitempty"`
	Achievements    *string `json:"achievements,omitempty"`
	Occupation      *string `json:"occupation,omitempty"`
}

// UpdatePersonRequest represents the request to update a person
type UpdatePersonRequest struct {
	FamilyName      *string `json:"family_name,omitempty"`
	GivenName       *string `json:"given_name,omitempty"`
	GenerationName  *string `json:"generation_name,omitempty"`
	CourtesyName    *string `json:"courtesy_name,omitempty"`
	ArtName         *string `json:"art_name,omitempty"`
	EnglishName     *string `json:"english_name,omitempty"`
	Gender          *string `json:"gender,omitempty" binding:"omitempty,oneof=M F U"`
	BirthDate       *string `json:"birth_date,omitempty"`
	BirthDateLunar  *string `json:"birth_date_lunar,omitempty"`
	BirthPlace      *string `json:"birth_place,omitempty"`
	DeathDate       *string `json:"death_date,omitempty"`
	DeathPlace      *string `json:"death_place,omitempty"`
	BurialPlace     *string `json:"burial_place,omitempty"`
	IsDeceased      *bool   `json:"is_deceased,omitempty"`
	IsAdopted       *bool   `json:"is_adopted,omitempty"`
	Biography       *string `json:"biography,omitempty"`
	Achievements    *string `json:"achievements,omitempty"`
	Occupation      *string `json:"occupation,omitempty"`
}

// PersonWithRelations includes related people
type PersonWithRelations struct {
	Person
	Parents       []Person        `json:"parents,omitempty"`
	Spouses       []SpouseInfo    `json:"spouses,omitempty"`
	Children      []ChildInfo     `json:"children,omitempty"`
	Siblings      []Person        `json:"siblings,omitempty"`
}

// SpouseInfo represents a spouse relationship
type SpouseInfo struct {
	Spouse        Person   `json:"spouse"`
	MarriageID    int64    `json:"marriage_id"`
	MarriageDate  *string  `json:"marriage_date,omitempty"`
	MarriageType  string   `json:"marriage_type"`
	EndDate       *string  `json:"end_date,omitempty"`
	EndReason     *string  `json:"end_reason,omitempty"`
}

// ChildInfo represents a child relationship
type ChildInfo struct {
	Child          Person   `json:"child"`
	RelationshipType string `json:"relationship_type"`
	BirthOrder     *int     `json:"birth_order,omitempty"`
	IsHeir         bool     `json:"is_heir"`
}
